File: Algoritmos de Ordenacao/java/Mergesort.java
import java.util.Random;

public class Mergesort {
    //variáveis globais
    private static int[] array = new int[10];
    public static int arraySize = array.length;

    private static void aleatorio() {
        Random random = new Random();
        for(int i = 0; i < arraySize; i++) {
            array[i] = random.nextInt(100);
        }
    }

    private static void intercalar(int esquerda, int meio, int direita) {
        //dados
        int nEsq = (meio + 1) - esquerda;
        int nDir = direita - meio;
        int iEsq = 0, iDir = 0, i;

        //criar subarrays
        int[] arrayEsq = new int[nEsq + 1];
        int[] arrayDir = new int[nDir + 1];

        arrayEsq[nEsq] = arrayDir[nDir] = 0x7FFFFFFF;

        //montar subarrays
        for (iEsq = 0; iEsq < nEsq; iEsq++) {
            arrayEsq[iEsq] = array [esquerda+iEsq];
        }

        for (iDir = 0; iDir < nDir; iDir++){
            arrayDir[iDir] = array[(meio+1)+iDir];
        }

        //Intercalacao propriamente dita
        for (iEsq = iDir = 0, i = esquerda; i <= direita; i++){
            array[i] = (arrayEsq[iEsq] <= arrayDir[iDir]) ? arrayEsq[iEsq++] : arrayDir[iDir++];
        }
    }

    public static void mergesort(int esquerda, int direita) {
        if(esquerda < direita) {
            int meio = (esquerda + direita) / 2;
            mergesort(esquerda, meio);
            mergesort(meio+1, direita);
            intercalar(esquerda, meio, direita);
        }
    }

    private static void mostrar() {
        System.out.print("[");

        for (int i = 0; i < arraySize; i++) {
            System.out.print(" "  + array[i]);
        }

        System.out.println(" ]");
    }

    public static void main(String[] args) {
        //criar novo array de randoms
        aleatorio();

        //mostrar random array
        System.out.println("Array desordenado: ");
        mostrar();

        //ordenar array
        mergesort(0, arraySize - 1);

        //mostrar array ordenado
        System.out.println("Array ordenado: ");
        mostrar();
    }
}

File: Algoritmos de Ordenacao/c/Mergesort.c
#include "stdio.h"
#include "stdlib.h"
#include "limits.h"

#define MAXTAM 10

//variavel globais
int array[MAXTAM];

//funcoes
void aleatorio();
void mergesort(int esq, int dir);
void intercalar(int esq, int meio, int dir);
void mostrar();

int main(void) {
    aleatorio();
    mostrar();
    mergesort(0, MAXTAM);
    mostrar();
}

void aleatorio() {
    for ( int i = 0; i < MAXTAM; i++) {
        array[i] = rand() % 100;
    }
}

void intercalar(int esq, int meio, int dir) {
    //dados
    int iEsq = 0, iDir = 0, i = 0;
    int nEsq = (meio + 1) - esq;
    int nDir = dir - meio;

    //criar subarrays
    int arrayEsq[nEsq + 1], arrayDir[nDir + 1];

    arrayEsq[nEsq] = arrayDir[nDir] = INT_MAX;

    //montar subarrays
    for (iEsq = 0; iEsq < nEsq; iEsq ++) {
        arrayEsq[iEsq] = array[esq+iEsq];
    }

    for (iDir = 0; iDir < nDir; iDir++) {
        arrayDir[iDir] = array[(meio + 1) + iDir];
    }

    //intercalação propriamente dita
    for (iEsq = iDir = 0, i = esq; i <= dir; i++){
        array[i] = (arrayEsq[iEsq] <= arrayDir[iDir]) ? arrayEsq[iEsq++] : arrayDir[iDir++];
    }
}

void mergesort(int esq, int dir) {
    if ( esq < dir ) {
        int meio = (esq + dir) / 2;
        mergesort(esq, meio);
        mergesort(meio+1, dir);
        intercalar(esq, meio, dir);
    }
}

void mostrar() {
    printf("%s", "[ ");
    for (int i = 0; i < MAXTAM; i++){
        printf("%i ", array[i]);
    }
    printf("%s\n", "]");
}